util: Improve documentation and negation of args

hodlinator opened the pull request:
- Document `-color` as only applying to `-getinfo`, to be less confusing for bitcoin-cli users.
- No longer print version information when getting passed `-noversion`.
- Disallow `-nodatadir` as we cannot run without one. It was previously interpreted as a mix of unset and as a relative path of "0".
- Support `-norpccookiefile`
- Support `-nopid`
- Properly support `-noconf` (instead of working by accident). Also detect when directories are specified instead of files.

Prompted by investigation in https://github.com/bitcoin/bitcoin/pull/16545#pullrequestreview-2316714013.

hodlinator committed:
args: Disallow -nodatadir

Does not make sense to run without a datadir.

Prior to this change it would be interpreted as a mix of unset and as a relative path of "0".

rkrux reviewed (commented):
Concept ACK 9e130811a3f68aea77ddeecf667aa8389a2940e8

Thanks for catching issues here, gave me an opportunity to read the app init code a bit more!

Successful make and functional tests. Left couple suggestions in the first 2 commits. Disallowing `noconf` and `nodatadir` conceptually makes sense to me. I tested with these 2 args and the output I received seems okay to me.

```
➜  bitcoin git:(2024/11/invalid_args) ✗ bitcoinclireg -noconf
Error parsing command line arguments: Negating of -conf is meaningless and therefore forbidden
➜  bitcoin git:(2024/11/invalid_args) ✗ bitcoinclireg -nodatadir
Error parsing command line arguments: Negating of -datadir is meaningless and therefore forbidden
```

Though I don't completely understand why this comment _"ok to not have a config file"_ (and subsequent flow) is allowed because when I ran the CLI without any arg I received the version info, help options, & too few args message.
https://github.com/hodlinator/bitcoin/blob/2024/11/invalid_args/src/common/config.cpp#L138

rkrux commented:
> Though I don't completely understand why this comment "ok to not have a config file" (and subsequent flow) is allowed

My guess it is applicable when the daemon is run without a conf because this portion lies in `common/`.

hodlinator committed:
doc args: Document narrow scope of -color

hodlinator commented:
> > Though I don't completely understand why this comment "ok to not have a config file" (and subsequent flow) is allowed
> 
> My guess it is applicable when the daemon is run without a conf because this portion lies in `common/`.

My interpretation is that we shouldn't trigger an error on not finding a configuration file in the specified or default location. If it's there we read it, otherwise just skip.

hodlinator commented on src/bitcoin-cli.cpp:
```diff
@@ -144,10 +144,10 @@ static int AppInitRPC(int argc, char* argv[])
         return EXIT_FAILURE;
     }
-    if (argc < 2 || HelpRequested(gArgs) || gArgs.IsArgSet("-version")) {
+    if (argc < 2 || HelpRequested(gArgs) || gArgs.GetBoolArg("-version", false)) {
```
I'm relying on this comment from unmerged #16545:
https://github.com/bitcoin/bitcoin/blob/b5ef85497436c3e9e60c760465d8991592efef07/src/common/args.h#L112-L114

One could imagine a user having a config file containing:
```
version=1
<...many lines...>
noversion=1
```
The last occurrence of a bool option takes precedence. Agree it feels very contrived for this specific setting, but I'm trying to stick to the general principle of trying to allow negation if possible.

hodlinator commented on src/bitcoin-cli.cpp:
```diff
@@ -144,10 +144,10 @@ static int AppInitRPC(int argc, char* argv[])
         std::string strUsage = CLIENT_NAME " RPC client version " + FormatFullVersion() + "\n";
 
-        if (gArgs.IsArgSet("-version")) {
+        if (gArgs.GetBoolArg("-version", false)) {
```
Done!

hodlinator commented:
> > Disallow -noconf since a config file is required
> 
> But it's not. Seems like `-noconf` should work to ignore the config file.

As mentioned in the commit message for 2d65e6ae0756b87e82dbafd032308e8687ec7941:

> ArgsManager::GetConfigFilePath() asserts that a path is always set. Explicitly negating it is therefore invalid."
>
> Previously, -noconf would result in an ifstream being opened to the ".bitcoin"-directory (not a file) in ArgsManager::ReadConfigFiles(), also circumventing the assert in aforementioned GetConfigFilePath().
> 
> (-conf="" is still allowed to reset to the default bitcoin.conf-file location)."

It's sheer luck when it comes to how `ifstream` treats opening of directories that has allowed this before the PR.

Here is `GetConfigFilePath`:

https://github.com/bitcoin/bitcoin/blob/2d65e6ae0756b87e82dbafd032308e8687ec7941/src/common/args.cpp#L761-L765

The part of `ArgsManager::ReadConfigFiles` which with `-noconf` would open a directory in the `ifstream`, and treat it as `.good()`.

https://github.com/bitcoin/bitcoin/blob/2d65e6ae0756b87e82dbafd032308e8687ec7941/src/common/config.cpp#L121-L140

`ReadConfigStream` calls into `GetConfigOptions` which gets as far as trying to `getline` from the directory `ifstream`, only to return true (success).

https://github.com/bitcoin/bitcoin/blob/2d65e6ae0756b87e82dbafd032308e8687ec7941/src/common/config.cpp#L33-L38

It would be good to add detection of "directory as config path" and trigger an error, regardless of which path is given. But I'm still inclined to disallow `-noconf` because of the assert in `GetConfigFilePath`.

l0rinc commented:
Concept ACK

hodlinator commented:
Added commit which now guards against passing directory paths as `-conf` values as per https://github.com/bitcoin/bitcoin/pull/31212#issuecomment-2468421728.

ryanofsky commented:
Code review d9cfb2a314caa1cdcae8e47a03dad557ad96e069.

This generally looks good except commit 2d65e6ae0756b87e82dbafd032308e8687ec7941 disallowing `-noconf`. Specifying `-noconf` to disable reading from the default config file path is useful and something that should be supported, even if now it's only working in a very quirky way by treating the datadir as an empty configuration file. I think a better implementation would not look too different though, maybe like:

```diff
--- a/src/common/config.cpp
+++ b/src/common/config.cpp
@@ -131,7 +131,7 @@ bool ArgsManager::ReadConfigFiles(std::string& error, bool ignore_invalid_keys)
     std::ifstream stream{conf_path};
 
     // not ok to have a config file specified that cannot be opened
-    if (IsArgSet("-conf") && !stream.good()) {
+    if ((IsArgSet("-conf") && !IsArgNegated("-conf")) && !stream.good()) {
         error = strprintf("specified config file \"%s\" could not be opened.", fs::PathToString(conf_path));
         return false;
     }
@@ -213,7 +213,7 @@ bool ArgsManager::ReadConfigFiles(std::string& error, bool ignore_invalid_keys)
 
 fs::path AbsPathForConfigVal(const ArgsManager& args, const fs::path& path, bool net_specific)
 {
-    if (path.is_absolute()) {
+    if (path.is_absolute() || path.empty()) {
         return path;
     }
     return fsbridge::AbsPathJoin(net_specific ? args.GetDataDirNet() : args.GetDataDirBase(), path);
```

I don't think it's right to infer from `Assert(m_config_path)` that a configuration file is always required by because that assert is really meant to catch initialization order bugs. The type of `m_config_path` is `std::optional<fs::path>` not `fs::path` so an empty path is ok there.

hodlinator committed:
args: Support -nopid

hodlinator commented:
@ryanofsky thanks for the review! Updated to cover negation for all users of `AbsPathForConfigVal`.

ryanofsky reviewed (commented):
Code review 77c8a538b7164f8b65576600eb4668ac160f6b14. Thanks for the updates! This all looks great, my only substantial comments are about the RPC commit which I think is more complicated than it needs to be and may have a bug where it may allow an empty auth string to be accepted if -norpccookiefile is specified, or if there is an initialization failure.

It may be nice to add python functional tests to exercise some of the new behaviors. It would be nontrivial work to add test coverage for all of these behaviors, but it doesn't have to be all-or-nothing. Just having some new end-to-end tests of argument parsing edge cases would be a nice improvement to our existing test coverage.

hodlinator reviewed (commented):
Thanks for continued review and reminding me to raise the bar for the PR @ryanofsky.

Added tests which helped uncover that *bitcoin-cli* might as well also support `-norpccookiefile` too, and part of *src/common/init.cpp* which was not handling `-noconf` correctly.

l0rinc commented:
I'd like to review this, but first, I'd like to understand why it is still in draft form.

hodlinator commented:
> I'd like to review this, but first, I'd like to understand why it is still in draft form.

Had a linefeed issue on Windows in the previous push (and was in a WIP state over the weekend, before that).. but pretty sure it's ready to go.

l0rinc commented on src/httprpc.cpp:
```diff
@@ -294,21 +296,26 @@ static bool InitRPCAuthentication()
 
-        if (!GenerateAuthCookie(&strRPCUserColonPass, cookie_perms)) {
-            return false;
+        if (std::optional<bool> result{GenerateAuthCookie(&strRPCUserColonPass, cookie_perms)}) {
```
You gotta' love these 3-state-booleans :)

nit: To reduce the noise, consider simplified to either
```suggestion
        if (std::optional result{GenerateAuthCookie(&strRPCUserColonPass, cookie_perms)}) {
```
or
```suggestion
        if (auto result{GenerateAuthCookie(&strRPCUserColonPass, cookie_perms)}) {
```

hodlinator replied:
I prefer documenting the type.

l0rinc replied:
it's quite noisy in this case, but don't have string feelings about it :)

l0rinc commented on test/functional/interface_bitcoin_cli.py:
in `test: -norpccookiefile`:

What's the reason for separating the test from the impl? I would understand if this were before the fix (we could check that it reproduces the desired behavior and fails - but we can't do that here since all tests should fail)

hodlinator replied:
I want it to be easy for reviewers to drop the impl commits or re-order and run what is in the test commits to verify that I'm actually fixing issues.

l0rinc replied:
I usually soft reset and revert the non-test changes to test - even simpler usually than an interactive rebase.
I still think the change and test should be in the same commit to give proper context.

l0rinc replied:
Given that currently the CI checks every commit, do we have a way to specifying to the CI that the test commit is meant to fail, i.e. demonstrates the need for the upcoming fix? Or would we add an ignored test before the fix and just unignore it with the fix? I still think that in the current setup it makes more sense to commit them as a singe work-unit - fixes and testing and benching and documenting them are not extras, they're part of the same work unit.

l0rinc replied:
I'm not against that, my concern is just that reproducing old behavior before the fix isn't very useful activity - it's a lot more useful to see that a test fails before the fix with the correct error (i.e TDD), but that's not really possible currently - unless we add something like https://github.com/spockframework/spock/blob/master/spock-core/src/main/java/spock/lang/PendingFeature.java#L13-L17 (where we invert the test, making sure it fails (that's when the test is green) and remove the annotation only in the fix commit).

ryanofsky replied:
> I'm not against that, my concern is just that reproducing old behavior before the fix isn't very useful activity

Strongly disagree :). In 2ca1460ae3a7217eaa8c5972515bf622bedadfce if tests had not been added *before* that commit checking for incorrect HTTP error codes, you would not know from looking at the change what the behavior was before the bugfix, or how new tests were connected to it. It is very useful to have test coverage for bugs before they are fixed.

> it's a lot more useful to see that a test fails before the fix with the correct error 

That's another benefit this approach provides. If the bugfix commit show a line of test code changing from `assert_equal(status, 200)` to `assert_equal(status, 400)` you know the check would have failed without the bugfix because `status` variable cannot be equal to two values at the same time. By contrast if the bugfix commit just included an entirely new `assert_equal(status, 400)`  check, it could tell you you nothing about whether the check would fail without the bugfix. It would only be possible determine that from manual testing or from a theoretical CI extension which our CI system does not implement.

Testing for bugs before fixing them is strictly better approach for making changes more understandable and providing more information about test coverage without requiring manual testing or theoretical CI extensions.

l0rinc replied:
> if tests had not been added before that commit checking for incorrect HTTP error codes, you would not know from looking at the change what the behavior was before the bugfix

The test would fail before the fix with the exact reason, not sure we're in a disagreement here.

> If the bugfix commit show a line of test code changing from [...]
> if the bugfix commit just included an entirely new [...]

But my point was exactly that if there's a failing test in the commit before the fix, which reproduces the error, you're not allowed to change the test in the commit that does the fix - to prove that you've solved the problem and not just castrated the test.
By modifying both the test and fix in the same commit, we don't know how the test would have behaved without the fix (or whether the previous test would pass with the fix). So the only reliable solution I see is adding failing tests before the fixing commit and not modifying them in the fix commit to make sure the exact same checks were done on both versions.
Same with benchmarks, I don't trust the performance comparison if the benchmarks have also changed when the code was changed.

l0rinc replied:
That's one way to do it, but in that case it's possible that the test passes before the fix as well (had many such reviews), which would indicate that we haven't actually reproduced the problem, therefore the fix might not do what it claims.
Only by having the exact same test fail before the fix and pass after (without any change to the test) can we be sure the fix had an effect on the test.

l0rinc replied:
Long story short: I would find it suspicious if the test of a specific fix passes without the fix.
The cleanest way I found to make sure this isn't the case is to add the test commit before the fix commit and make sure CI fails if the test passes without the fix (see mentioned Spock [PendingFeature](https://github.com/spockframework/spock/blob/master/spock-core/src/main/java/spock/lang/PendingFeature.java#L13-L17)).
But since we don't have that here, I would prefer having the test in the same work unit as the fix, to provide the necessary context for the current fix.
Is this still controversial?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31212#discussion_r1854660293

> But since we don't have that here, I would prefer having the test in the same work unit as the fix, to provide the necessary context for the current fix.
Is this still controversial?

I think it's a perfectly reasonable suggestion. To paraphrase, you are saying that when code changes are made in a commit without test changes (which is how the PR is currently structured, though not what I was suggesting) then it is harder for reviewers and git blame users to determine what the associated tests are. Also including test changes in code commits makes it easier to write an automated checker that could determine whether the test changes would have passed without the code changes, and would indicate how meaningful the test coverage is. 

These are reasonable things to want, just like having smaller more focused commits is a reasonable thing to want. So your approach and hodlinators's approach are both fine choices AFAICT. The only thing I'd add is that my suggestion reduces the tradeoff between these approach and gives the best of both worlds: including tests in the same commits as code changes, but minimizing the test diffs there, and additionally making it more obvious what differences in old and new behavior are.

hodlinator replied:
To me this is a good pattern:
1. Document former *correct behavior* by adding tests.
2. Change implementation and tests to conform to *new behavior*.

However, if 1) is about documenting *buggy behavior*, committing a test for that in a way that makes them succeed feels off to me. In that case I would prefer adding tests that *prove failure* with some special annotation in the commit message for CI to expect failure like what @l0rinc [linked here (Spock)](https://github.com/bitcoin/bitcoin/pull/31212#discussion_r1854660293).

I'm probably biased from having worked in other workflows which encouraged less granular commits.

l0rinc commented on src/common/config.cpp:
```diff
@@ -128,13 +129,21 @@ bool ArgsManager::ReadConfigFiles(std::string& error, bool ignore_invalid_keys)
-    if (IsArgSet("-conf") && !stream.good()) {
-        error = strprintf("specified config file \"%s\" could not be opened.", fs::PathToString(conf_path));
-        return false;
+        // not ok to have a config file specified that cannot be opened
```
nit: double negation makes it more complicated than need be. If you insist on having a comment:
```suggestion
        // Config file must be accessible
```

hodlinator replied:
Good point about double negation. Changed to:
```C++
// If the file is explicitly specified, it must be readable
```
since the distinction is important.

l0rinc commented on src/init/common.cpp:
```diff
@@ -125,8 +125,10 @@ bool StartLogging(const ArgsManager& args)
-        // Warn if no conf file exists at path provided by user
-        InitWarning(strprintf(_("The specified config file %s does not exist"), fs::PathToString(config_file_path)));
+        if (!args.IsArgNegated("-conf")) {
+            // Warn if no conf file exists at path provided by user
```
not sure what this comments adds that the code doesn't already explain

l0rinc commented on test/functional/feature_config_args.py:
in `test: Add tests for directory in place of config files, -noconf`:

Same as with the other test, I think it belongs with the related modification

hodlinator replied:
https://github.com/bitcoin/bitcoin/pull/31212#discussion_r1851093183

l0rinc reviewed (commented):
I love that you search for these inconsistencies and attack them head on.
+1 for adding representative tests (which I didn't review in detail yet, will do in the next round of reviews).

hodlinator commented on src/bitcoin-cli.cpp:
```diff
@@ -94,7 +94,7 @@ static void SetupCliArgs(ArgsManager& argsman)
 
     SetupChainParamsBaseOptions(argsman);
-    argsman.AddArg("-color=<when>", strprintf("Color setting for CLI output (default: %s). Valid values: always, auto (add color codes when standard output is connected to a terminal and OS is not WIN32), never.", DEFAULT_COLOR_SETTING), ArgsManager::ALLOW_ANY | ArgsManager::DISALLOW_NEGATION, OptionsCategory::OPTIONS);
+    argsman.AddArg("-color=<when>", strprintf("Color setting for CLI output (default: %s). Valid values: always, auto (add color codes when standard output is connected to a terminal and OS is not WIN32), never. Only applies to the output of -getinfo.", DEFAULT_COLOR_SETTING), ArgsManager::ALLOW_ANY | ArgsManager::DISALLOW_NEGATION, OptionsCategory::OPTIONS);
```
> Or is the `ArgsManager::DISALLOW_NEGATION` meant to discourage people from requesting that :D?

Indeed, although I agree with `-nocolor` potentially equaling `-color=never`, it is flagged as forbidden and therefore the process is stopped before reaching the quoted code. I'm a bit hesitant to overrule the prior intent of disallowing it.

hodlinator committed:
scripted-diff: Avoid printing version information for -noversion

-BEGIN VERIFY SCRIPT-
sed -i --regexp-extended 's/\b(gArgs|args)\.IsArgSet\("-version"\)/\1.GetBoolArg("-version", false)/g' $(git grep -l '-version')
-END VERIFY SCRIPT-

hodlinator commented on src/bitcoin-tx.cpp:
```diff
@@ -105,11 +105,11 @@ static int AppInitRawTx(int argc, char* argv[])
     fCreateBlank = gArgs.GetBoolArg("-create", false);
 
-    if (argc < 2 || HelpRequested(gArgs) || gArgs.IsArgSet("-version")) {
+    if (argc < 2 || HelpRequested(gArgs) || gArgs.GetBoolArg("-version", false)) {
```
`IsArgSet` would be true when the user passes `-noversion`, so the `if`-block would be entered, which is intended to be avoided in this PR.

hodlinator replied:
`-version=true` and `-version=false` are both interpreted as `-version=0` (there is no special parsing for booleans expressed as strings).
`-version=0` is equivalent to `-noversion` or `-noversion=1`.

`IsArgSet()` returns `true` if the user has set a given arg to *any value* (including negation), through command line or settings-file.

l0rinc replied:
The part I'm not sure about is checking `gArgs.GetBoolArg("-version", false)` twice - since the inner check has an else, which is a bit confusing, since that can only happen if `argc < 2 || HelpRequested(gArgs)` are true - seems like the condition can be simplified, especially since `if (argc < 2) {` is rechecked later. (please verify for other such conditions, not necessarily this one)

hodlinator commented on src/common/config.cpp:
I fail to see the typo, mind pointing it out?

hodlinator commented on src/init.cpp:
```diff
@@ -483,7 +483,7 @@ void SetupServerArgs(ArgsManager& argsman, bool can_listen_ipc)
     argsman.AddArg("-coinstatsindex", strprintf("Maintain coinstats index used by the gettxoutsetinfo RPC (default: %u)", DEFAULT_COINSTATSINDEX), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     argsman.AddArg("-conf=<file>", strprintf("Specify path to read-only configuration file. Relative paths will be prefixed by datadir location (only useable from command line, not configuration file) (default: %s)", BITCOIN_CONF_FILENAME), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
-    argsman.AddArg("-datadir=<dir>", "Specify data directory", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    argsman.AddArg("-datadir=<dir>", "Specify data directory", ArgsManager::ALLOW_ANY | ArgsManager::DISALLOW_NEGATION, OptionsCategory::OPTIONS);
```
Potentially yes, but I'm trying to keep the scope of the PR to the args I discovered here: https://github.com/bitcoin/bitcoin/pull/16545#pullrequestreview-2316714013

Open to exploring it in a follow-up!

hodlinator commented on src/httprpc.cpp:
```diff
@@ -294,21 +296,26 @@ static bool InitRPCAuthentication()
+            LogInfo("RPC authentication cookie file generation is disabled.");
         }
     } else {
         LogPrintf("Config options rpcuser and rpcpassword will soon be deprecated. Locally-run instances may remove rpcuser to use cookie-based auth, or may be replaced with rpcauth. Please see share/rpcauth for rpcauth auth generation.\n");
```
You mean
```
LogInfo("Using random cookie authentication.\n");
```
was run independently of whether `GenerateAuthCookie` returned `true` or `false`, but now it may not be? It seemed wrong to me so I made it only happen when the cookie was actually generated.

If you mean that
```
LogPrintf("Config options rpcuser and rpcpassword will soon be deprecated. Locally-run instances may remove rpcuser to use cookie-based auth, or may be replaced with rpcauth. Please see share/rpcauth for rpcauth auth generation.\n");
```
was emitted when `GenerateAuthCookie` was `true`, I think you may be mixing up some braces/indentation.

l0rinc replied:
Yes, I was confused by indentation, resolve the comment por favor

hodlinator commented on src/httprpc.cpp:
```diff
@@ -148,7 +149,8 @@ static bool RPCAuthorized(const std::string& strAuth, std::string& strAuthUserna
 
     //Check if authorized under single-user field
-    if (TimingResistantEqual(strUserPass, strRPCUserColonPass)) {
+    // strRPCUserColonPass is empty when -norpccookiefile is specified
```
Yeah, was itching to do that too, thanks for a second vote. :)

ryanofsky reviewed (approved):
Code review ACK 2bf7e59776fcfb2c8a477cf2abb59606f7b4aacc. This is a nice improvement over status quo, but I'm pretty sure two of the commits 21bc3a3657fea5ed44172963639ee8ca5cd8e46c and 879a8d3a6820a86e6faa8d91f92d0cf374700fb8 are more complicated than they need to be and less robust than they could be.

I posted a simplified [branch](https://github.com/ryanofsky/bitcoin/commits/review.31212.4-edit) ([diff](https://github.com/ryanofsky/bitcoin/compare/review.31212.4..review.31212.4-edit)) and would recommend replacing:

- 21bc3a3657fea5ed44172963639ee8ca5cd8e46c with 43ec65a6bd5786a47fba91eec96cb68232104904
- 879a8d3a6820a86e6faa8d91f92d0cf374700fb8 with 3a856c86685a5ab6f3cb78b17cb52cbb1b9ad95b
- 2bf7e59776fcfb2c8a477cf2abb59606f7b4aacc with ebadf80f2fed1fc3513e76a48696e4f561ce577d (to fix tests)

Issues with current commit "args: Support -norpccookiefile for bitcoind and bitcoin-cli" (21bc3a3657fea5ed44172963639ee8ca5cd8e46c) are:

- Tristate optional\<bool> is unnecessary when the function can just return true in successful states and false in failed ones.

- New "Unable to set specified permissions on cookie file as we aren't generating one" warning doesn't seem like a good idea. Warning about valid but unused options is not consistent with how most other options are handled, and having a warning for this niche case makes the code harder to follow. Also the warning is not shown reliably, such as when when -rpcpassword is specified and -rpccookieperms is also ignored.

- New `g_http_rpc_auth_initialized` variable is unnecessary. Its only purpose to preserve an old "Belt-and-suspenders" check but there is nothing particularly useful about that check. The check made sense previously to detect the formerly invalid state of the "user:pass" string being empty. But now that -norpccookiefile works, it makes sense for that string to be empty, and doesn't make sense to add a *new* invalid state just for the sake of having some invalid state. At least I can think of more ways adding an unnecessary invalid state could cause bugs than it could prevent them.

- New "RPC authentication cookie file reading is disabled" debug log in bitcoin-cli code doesn't seem useful. AFAICT bitcoin-cli doesn't show debug messages and this message doesn't add much information beyond what the more actionable `bitcoin-cli` error message already conveys. Having this doesn't seem worth hit to complexity & readability of the GetAuthCookie function, and if the error message is unclear it would be better to just improve the error message.

- Very subjective, but I think the unrelated code cleanups in this commit make it hard to follow and are very distracting. Would be much happier to see them dropped or moved to a refactoring commit.

Issues with current commit "args: Properly support -noconf and catch directories" (879a8d3a6820a86e6faa8d91f92d0cf374700fb8) are:

- The `fs::is_directory` checks seem to be covering up a problem rather than fixing it. AFAICT the problem with `ReadConfigFiles` is that is reading all the config files without bothering to check that the reads are successful. If the code changed to check `ifstream::eof` and verify that the files were read successfully it would be simpler and more reliable.

- The "included config file \"%s\" is a directory." error formatting is buggy. It references the path of the main config file not the included file.

- The "data directory contains a config file which is ignored because -noconf is specified error" seems to introduce an inconsistency with how -noconf is handled. I would expect -noconf to always ignore the configuration file, not sometimes ignore it and other times show an error an error about it, and for which of these happens to depend on how the data directory was located.

- The new IsArgNegated call in StartLogging seems to be modifying dead code. Or I don't see how the "The specified config file %s does not exist" warning would be triggered when there is already an error about that. Seems better to just drop the redundant code here.

l0rinc commented on test/functional/rpc_users.py:
```diff
@@ -166,11 +189,18 @@ def run_test(self):
+        node0_cookie_file_path_tmp = self.nodes[0].chain_path / ".cookie.tmp"
+        node0_cookie_file_path_tmp.mkdir()
         self.nodes[0].assert_start_raises_init_error(expected_msg=init_error)
+        node0_cookie_file_path_tmp.rmdir()
```
will the `node0_cookie_file_path_tmp` file be kept when assertion fails on purpose? If not, you could wrap it in a try/finally (deleting regardless of the assertion's outcome)

hodlinator replied:
Tests are running in a scratch directory - if part of the test fails, we don't worry about leaving some files behind.

l0rinc replied:
Can we remove the cleanup in that case?

hodlinator replied:
The cleanup is still there so that other tests following it in the Python script are not impacted.

l0rinc replied:
so do I understand it correctly that it's fine for other tests to be affected when this one already fails, leaving the directory?

l0rinc commented on src/rpc/request.h:
```diff
@@ -24,7 +24,7 @@ UniValue JSONRPCReplyObj(UniValue result, UniValue error, std::optional<UniValue
 
 /** Generate a new RPC authentication cookie and write it to disk */
-bool GenerateAuthCookie(std::string* cookie_out, std::optional<fs::perms> cookie_perms=std::nullopt);
+std::optional<bool> GenerateAuthCookie(std::string* cookie_out, std::optional<fs::perms> cookie_perms=std::nullopt);
```
as mentioned before, I find this tri-state-boolean quite confusing - if 3 states are indeed needed, maybe it's hiding a different structure such as a variant or enum or `optional<Error>`

hodlinator replied:
Reverted in later pushes.

l0rinc commented on src/rpc/request.cpp:
```diff
@@ -142,6 +150,10 @@ bool GetAuthCookie(std::string *cookie_out)
     fs::path filepath = GetAuthCookieFile();
+    if (filepath.empty()) {
+        LogDebug(BCLog::RPC, "RPC authentication cookie file reading is disabled.");
+        return false;
```
why is this false and not an empty optional? As mentioned before, I don't really see the line

hodlinator replied:
`GetAuthCookieFile()` internally uses `GetPathArg()` and follows it's pattern (predating this PR):
https://github.com/bitcoin/bitcoin/blob/7590e93bc73b3bbac641f05d490fd5c984156b33/src/common/args.cpp#L272-L274

But might change to `optional` if you insist. Tried out changing it in 01fa9d41d12a68f88f6251ae43234dd250a53bd5. Works fairly well, even uncovered a minor pre-existing bug in that if `fs::permissions` fails, we currently print out the wrong filename. With that bug it feels like it could arguably be its own micro-PR.

l0rinc replied:
I'm fine with doing it in a different PR - will leave it up to you

l0rinc commented on src/qt/bitcoin.cpp:
```diff
@@ -582,8 +582,8 @@ int GuiMain(int argc, char* argv[])
-    if (HelpRequested(gArgs) || gArgs.IsArgSet("-version")) {
-        HelpMessageDialog help(nullptr, gArgs.IsArgSet("-version"));
+    if (HelpRequested(gArgs) || gArgs.GetBoolArg("-version", false)) {
+        HelpMessageDialog help(nullptr, gArgs.GetBoolArg("-version", false));
```
What's the reason for not wanting an `about` window regardless of the `-version` flag's value? Doesn't `gArgs.IsArgSet("-version")` make more sense here?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31212#discussion_r1853683566

> What's the reason for not wanting an `about` window regardless of the `-version` flag's value? Doesn't `gArgs.IsArgSet("-version")` make more sense here?

The idea is for -noversion and -version=0 to be the same as not specifying a -version argument. -version is just a binary option, it should choose between 2 behaviors, not 3

l0rinc replied:
what about `-noversion=0` :p

There has to be a better way, this is so confusing :/

l0rinc commented on src/init/common.cpp:
```diff
@@ -125,8 +125,9 @@ bool StartLogging(const ArgsManager& args)
     if (fs::exists(config_file_path)) {
         LogPrintf("Config file: %s\n", fs::PathToString(config_file_path));
     } else if (args.IsArgSet("-conf")) {
```
given that checking the args is not for free (triggers locks and merges from multiple sources), maybe we could introduce a `IsArgEnabled` which gets rid of this confusing combination (where `-noconf=0` would be set and not negated, I think)

hodlinator committed:
test: Harden testing of cookie file existence

hodlinator committed:
test: -norpccookiefile

Both bitcoind and bitcoin-cli.

hodlinator committed:
test: Add tests for directories in place of config files

hodlinator commented:
Pushed 51e34ca022d6704dcf1fa955e3764a97d91bbba8

---

Unfortunately I wasn't able to use the simplification suggested by @ryanofsky:
https://github.com/bitcoin/bitcoin/blob/3a856c86685a5ab6f3cb78b17cb52cbb1b9ad95b/src/common/config.cpp#L138-L141

When opening a directory with an `ifstream`, Windows CI would not return `true` for `is_open()`:
https://github.com/hodlinator/bitcoin/actions/runs/12052081230/job/33604594555#step:12:3324

When opening a directory with an `ifstream` on MacOS CI, `is_open()` would return `true`, but the EOF-bit would also be set:
https://github.com/hodlinator/bitcoin/actions/runs/12052081230/job/33604593738#step:7:6158

So reverted to using `is_directory()` as it is more robust and clear.

---

\+ Commit 6305f701dee90fb85a72f0d5ae3cdcd57871008e - Improves *combine_logs.py* which helped me track down that I had forgotten to pass `-regtest` in a couple of calls to `assert_start_raises_init_error()` (creating mainnet *debug.log*s).

\+ Commit daa119d580d47be4c22a762b732ef80a570b0f08 - Refactors *feature_config_args.py* towards starting the node at beginning of test functions and stopping it at the end to reduce implicit dependencies.

Broke up `-noconf` and directory checking changes and tests into separate commits.

Simplified warning about ignored config file in *common/init.cpp*.

Added special output for config file being a directory in *init/common.cpp*.

l0rinc commented on src/rpc/request.cpp:
```diff
@@ -142,6 +149,9 @@ bool GetAuthCookie(std::string *cookie_out)
     std::string cookie;
     fs::path filepath = GetAuthCookieFile();
+    if (filepath.empty()) {
+        return true; // -norpccookiefile was specified
```
Do I understand correctly that we're basically treating an empty path as an `std::nullopt` here (and commenting it everywhere since we admit it's hard to understand)?
I understand that we can't fix everything in the same PR, but would like us to consider if it still makes sense to build on previous hacks.

l0rinc commented on src/init.cpp:
```diff
@@ -175,6 +175,8 @@ static fs::path GetPidFile(const ArgsManager& args)
 
 [[nodiscard]] static bool CreatePidFile(const ArgsManager& args)
 {
+    if (args.IsArgNegated("-pid")) return true;
```
it's hard to review 14 commits, as hinted before, could the fix be in the same commit as the test - I personally am lacking a lot of context and wouldn't want this commit to be merged without making sure a test is added as well

hodlinator replied:
PR does not contain a test for this as I thought it was straightforward enough to skip and I didn't want to add more things to review/maintain. If we pass `-nopid` we simply do not:
- Create the file
- Write the PID
- Set the `bool` to remember deleting the file
- (Delete the file)

https://github.com/bitcoin/bitcoin/blob/95a0104f2e9869799db84add108ae8c57b56d360/src/init.cpp#L165-L192

l0rinc commented on src/httprpc.cpp:
```diff
@@ -307,9 +304,11 @@ static bool InitRPCAuthentication()
         if (!GenerateAuthCookie(&strRPCUserColonPass, cookie_perms)) {
             return false;
         }
+        if (!strRPCUserColonPass.empty()) LogInfo("Using random cookie authentication.");
```
what if it's empty?

hodlinator replied:
Moved back the log message in latest push to clarify.

l0rinc commented on src/common/config.cpp:
in commit message `args: Properly support -noconf ...`:

> have been updated earlier in this PR (https://github.com/bitcoin/bitcoin/commit/bffd92f00f5bfbb5a622d0bd20bfeed9f8b10928 and https://github.com/bitcoin/bitcoin/commit/f5a14f40e5e2ad3504d91ef03929898b3cab85b4).

The commits [have changed since](https://github.com/bitcoin/bitcoin/pull/31212/commits/d7ecacf1dcce3b405cd35cffa1301067f713acce), but instead of updating, I would rather recommend removing hashes from the commit message

hodlinator replied:
Cheers, replaced hashes with commit message titles.

l0rinc commented on src/common/config.cpp:
```diff
@@ -128,12 +128,14 @@ bool ArgsManager::ReadConfigFiles(std::string& error, bool ignore_invalid_keys)
-        error = strprintf("specified config file \"%s\" could not be opened.", fs::PathToString(conf_path));
-        return false;
+    std::ifstream stream;
+    if (!conf_path.empty()) { // path is empty when -noconf is specified
```
This bothers me more than it should: we're adding comments instead of solving these problems with code - conf_path should be an optional in that case instead of coming up with special value

hodlinator replied:
We are currently using the nullness of `m_config_path` to signal something different, initialized/not, although I think it should probably be more explicit. See https://github.com/bitcoin/bitcoin/pull/31212#issuecomment-2470510859, and this code:

https://github.com/bitcoin/bitcoin/blob/2ac863bdf974d7b4b4e8314876d02270400b6240/src/common/args.cpp#L761-L772

We could possibly do `optional<optional<path>> m_config_path` to flag both initialization and negation... but I'm not sure I want to go down that road. Do you have any concrete suggestion?

l0rinc replied:
optional of optional sounds like a `std::variant` - but I'm fine with doing these in a separate PR, of course

l0rinc reviewed (changes requested):
Thanks for enduring through the struggles the Windows gods bestow upon you.
While there are parts I can't meaningfully review yet (no idea what to do with the cookies yet), I left a few comments about how some restructuring could help me understand the context better: if a commit should not be merged without another one, I would prefer combining them into a single commit - to avoid the bloat, it makes review harder.

hodlinator committed:
args: Catch directories in place of config files

Previously passing a directory path as -conf would lead to an ifstream being opened for it, and would not trigger any errors.

hodlinator committed:
logs: Use correct path and more appropriate macros in cookie-related code

filepath_tmp -> filepath in last message.

More material changes to nearby code in next commit.

ryanofsky commented on src/common/config.cpp:
```diff
@@ -130,6 +131,10 @@ bool ArgsManager::ReadConfigFiles(std::string& error, bool ignore_invalid_keys)
     const auto conf_path{GetConfigFilePath()};
     std::ifstream stream;
     if (!conf_path.empty()) { // path is empty when -noconf is specified
+        if (fs::is_directory(conf_path)) {
```
In commit "args: Catch directories in place of config files" (01b174493ec210c1c627083dfa9e9b86035c7eef)

If macos is setting the eof bit when using ifstream to read from a directory paths as described https://github.com/bitcoin/bitcoin/pull/31212#issuecomment-2506175947, treating directories like empty files, maybe we really do need these is_directory calls, since there might be no other way on macos to distinguish directories from empty files, though I suspect it probably is indicating some error with `stream.bad()` and we don't really need to do this.

The windows behavior described https://github.com/bitcoin/bitcoin/pull/31212#issuecomment-2506175947 of refusing to open directories is less surprising than the mac behavior, and could be worked around by using `fs::exists` instead of `is_open` as a more direct way of determine whether the path exists.

Overall, the current approach in this commit seems ok, but fragile to only check if the file paths are directories and not actually check that files were read successfully. I think a more ideal version of this taking windows and mac behavior into account would look like the following:

<details><summary>diff</summary>
<p>

```diff
--- a/src/common/config.cpp
+++ b/src/common/config.cpp
@@ -129,24 +129,19 @@ bool ArgsManager::ReadConfigFiles(std::string& error, bool ignore_invalid_keys)
     }
 
     const auto conf_path{GetConfigFilePath()};
-    std::ifstream stream;
-    if (!conf_path.empty()) { // path is empty when -noconf is specified
-        if (fs::is_directory(conf_path)) {
-            error = strprintf("Config file \"%s\" is a directory.", fs::PathToString(conf_path));
-            return false;
-        }
-        stream = std::ifstream{conf_path};
-        // If the file is explicitly specified, it must be readable
-        if (IsArgSet("-conf") && !stream.good()) {
-            error = strprintf("specified config file \"%s\" could not be opened.", fs::PathToString(conf_path));
-            return false;
-        }
-    }
-    // ok to not have a config file
-    if (stream.good()) {
+    // Don't try to read the config file if -noconf is specified (so
+    // conf_path.empty() is true), or if the -conf option is unset and the
+    // default config file path doesn't exist.
+    if (!conf_path.empty() && (IsArgSet("-conf") || fs::exists(conf_path))) {
+        std::ifstream stream{conf_path};
         if (!ReadConfigStream(stream, fs::PathToString(conf_path), error, ignore_invalid_keys)) {
             return false;
         }
+        // Error if config stream could not be read in full.
+        if (!stream.eof() || stream.bad()) {
+            error = strprintf("config file \"%s\" could not be read.", fs::PathToString(conf_path));
+            return false;
+        }
         // `-includeconf` cannot be included in the command line arguments except
         // as `-noincludeconf` (which indicates that no included conf file should be used).
         bool use_conf_file{true};
@@ -182,16 +177,11 @@ bool ArgsManager::ReadConfigFiles(std::string& error, bool ignore_invalid_keys)
             const size_t default_includes = add_includes({});
 
             for (const std::string& conf_file_name : conf_file_names) {
-                const auto include_conf_path{AbsPathForConfigVal(*this, fs::PathFromString(conf_file_name), /*net_specific=*/false)};
-                if (fs::is_directory(include_conf_path)) {
-                    error = strprintf("Included config file \"%s\" is a directory.", fs::PathToString(include_conf_path));
+                std::ifstream conf_file_stream{AbsPathForConfigVal(*this, fs::PathFromString(conf_file_name), /*net_specific=*/false)};
+                if (!ReadConfigStream(conf_file_stream, conf_file_name, error, ignore_invalid_keys)) {
                     return false;
                 }
-                std::ifstream conf_file_stream{include_conf_path};
-                if (conf_file_stream.good()) {
-                    if (!ReadConfigStream(conf_file_stream, conf_file_name, error, ignore_invalid_keys)) {
-                        return false;
-                    }
+                if (conf_file_stream.eof() && !conf_file_stream.bad()) {
                     LogPrintf("Included configuration file %s\n", conf_file_name);
                 } else {
                     error = "Failed to include configuration file " + conf_file_name;
--- a/test/functional/feature_config_args.py
+++ b/test/functional/feature_config_args.py
@@ -45,7 +45,7 @@ class ConfArgsTest(BitcoinTestFramework):
         self.stop_node(0)
         self.nodes[0].assert_start_raises_init_error(
             extra_args=['-regtest'],
-            expected_msg=f'Error: Error reading configuration file: Config file "{conf_path}" is a directory.',
+            expected_msg=f'Error: Error reading configuration file: config file "{conf_path}" could not be read.',
         )
         conf_path.rmdir()
         os.rename(conf_path.with_suffix('.confbkp'), conf_path)
@@ -55,7 +55,7 @@ class ConfArgsTest(BitcoinTestFramework):
             conf.write(f'includeconf={self.nodes[0].datadir_path}\n')
         self.nodes[0].assert_start_raises_init_error(
             extra_args=['-regtest'],
-            expected_msg=f'Error: Error reading configuration file: Included config file "{self.nodes[0].datadir_path}" is a directory.',
+            expected_msg=f'Error: Error reading configuration file: Failed to include configuration file {self.nodes[0].datadir_path}',
         )
 
         self.nodes[0].replace_in_config([(f'includeconf={self.nodes[0].datadir_path}', '')])
@@ -521,7 +521,7 @@ class ConfArgsTest(BitcoinTestFramework):
 
         # Check that an explicitly specified config file that cannot be opened fails
         none_existent_conf_file = default_data_dir / "none_existent_bitcoin.conf"
-        self.nodes[0].assert_start_raises_init_error(['-conf=' + f'{none_existent_conf_file}'], 'Error: Error reading configuration file: specified config file "' + f'{none_existent_conf_file}' + '" could not be opened.')
+        self.nodes[0].assert_start_raises_init_error(['-conf=' + f'{none_existent_conf_file}'], 'Error: Error reading configuration file: config file "' + f'{none_existent_conf_file}' + '" could not be read.')
 
         # Create the directory and ensure the config file now works
         new_data_dir.mkdir()
```
</p>
</details>

But current code seems fine and this could be done in a followup. Thanks for looking into this and making the current fix easy to understand.

hodlinator replied:
Would rather not spend more cycles on this when C++ appears to not specify how `ifstream` interacts with directories. Even if we were to get it working with your approach on the 3 platforms on current CI, others might get it compiling on other setups where `is_directory()` would again have been more robust.

I do see value in catching whether reading a file fails, but prefer keeping it out of scope of this PR.

ryanofsky reviewed (approved):
Code review ACK c972e691d937f49d7d3f44d2b19b4035c6e67aaf. Thanks for breaking the commits up make making things easy to understand.

Changes since last review were simplifying the -norpccookiefile and -noconf fixes by breaking them up and changing some implementation details. There are also new combine_logs and feature_config args test improvements.

hodlinator committed:
args: Properly support -noconf

-noconf would previously lead to an ifstream "successfully" being opened to the ".bitcoin"-directory (not a file). (Guards against the general case of directories as configs are added in grandchild commit to this one).

Other users of AbsPathForConfigVal() in combination with negated args have been updated earlier in this PR ("args: Support -nopid" and "args: Support -norpccookiefile...").

hodlinator committed:
args: Support -norpccookiefile for bitcoind and bitcoin-cli

Replaces belt & suspenders check for initialization in RPCAuthorized() with not allowing empty passwords further down.

hodlinator committed:
test refactor: feature_config_args.py - Stop nodes at the end of tests, not at the beginning

This ensures we don't needlessly start the node, and reduces implicit dependencies between test functions.

test_seed_peers() - Move assert calling RPC to verify correct chain after our own function actually started the node.

hodlinator committed:
test: combine_logs.py - Output debug.log paths on error

hodlinator committed:
test: Add tests for -noconf

hodlinator commented on src/init/common.cpp:
```diff
@@ -122,10 +123,13 @@ bool StartLogging(const ArgsManager& args)
     // Only log conf file usage message if conf file actually exists.
     fs::path config_file_path = args.GetConfigFilePath();
-    if (fs::exists(config_file_path)) {
+    if (args.IsArgSet("-conf") && args.IsArgNegated("-conf")) {
```
Thanks! Fixed in latest push.

hodlinator commented on src/init/common.cpp:
```diff
@@ -122,10 +123,13 @@ bool StartLogging(const ArgsManager& args)
         LogPrintf("Config file: %s\n", fs::PathToString(config_file_path));
     } else if (args.IsArgSet("-conf")) {
-        // Warn if no conf file exists at path provided by user
         InitWarning(strprintf(_("The specified config file %s does not exist"), fs::PathToString(config_file_path)));
```
I prefer having a "Config file: "-prefix to grep for in logs, so following that pattern for added messages here in this PR (but not going to touch the InitWarning as it might be shown in the GUI).

hodlinator commented on src/rpc/request.cpp:
```diff
@@ -106,24 +109,28 @@ bool GenerateAuthCookie(std::string* cookie_out, std::optional<fs::perms> cookie
     fs::path filepath = GetAuthCookieFile(false);
     if (!RenameOver(filepath_tmp, filepath)) {
-        LogInfo("Unable to rename cookie authentication file %s to %s\n", fs::PathToString(filepath_tmp), fs::PathToString(filepath));
+        LogWarning("Unable to rename cookie authentication file %s to %s\n", fs::PathToString(filepath_tmp), fs::PathToString(filepath));
```
Would rather not change to quote paths in all log messages of this function as it would touch even more lines.

hodlinator commented on test/functional/combine_logs.py:
```diff
@@ -79,11 +79,11 @@ def read_logs(tmp_dir):
+    debug_logs = list(pathlib.Path(tmp_dir).glob('node0/**/debug.log'))
+    if len(debug_logs) > 0:
+        path = debug_logs[0]
+        assert len(debug_logs) < 2, 'More than one debug.log, should never happen:\n\t' + '\n\t'.join([str(f) for f in debug_logs])
```
Thanks, agree that it's a bit redundant in an assert, leftover from previous comment. Updated.

hodlinator commented on test/functional/combine_logs.py:
```diff
@@ -79,11 +79,11 @@ def read_logs(tmp_dir):
-        assert next(glob, None) is None #  more than one debug.log, should never happen
+    # Find out what the folder is called that holds node 0's debug.log file
+    debug_logs = list(pathlib.Path(tmp_dir).glob('node0/**/debug.log'))
+    if len(debug_logs) > 0:
```
Current way is clearer & safer to me.

hodlinator commented on test/functional/feature_config_args.py:
```diff
@@ -27,9 +27,71 @@ def set_test_params(self):
         self.wallet_names = []
         self.disable_autoconnect = False
 
+    def setup_network(self):
```
I've updated that commit to better document that I'm overriding the 2 methods (unfortunately our *.python-version* is 3.10.14, python 3.12 includes `@override`).

hodlinator commented on test/functional/feature_config_args.py:
```diff
@@ -196,10 +257,10 @@ def test_args_log(self):
                 '-rpcuser=secret-rpcuser',
                 '-torpassword=secret-torpassword',
             ])
+        # Leave node running for test_seed_peers()
```
Updated commit to make this function also call `self.stop_node(0)`, hammering down the last nail that stood out. :)

Hopefully that makes it easier to review. The commit was already making test run order matter less, but the new version takes it one step further. There may still be other state than node stopped/running that prevents re-ordering, but that was the main thing I ran into when adding test functions in the later commits.

hodlinator commented on src/common/init.cpp:
```diff
@@ -65,26 +65,33 @@ std::optional<ConfigError> InitConfig(ArgsManager& args, SettingsAbortFn setting
-                return ConfigError{ConfigStatus::FAILED, Untranslated(error)};
+        if (fs::exists(base_config_path)) {
+            if (orig_config_path.empty()) {
+                LogWarning(
```
Having a latent *bitcoin.conf* is not ideal and something the user should correct if they want to run with `-noconf` long-term to reduce potential for confusion ("Why is my change to *bitcoin.conf* not getting picked up?!!" until realizing `-noconf` was used). But changed to `LogInfo` for now.

ryanofsky reviewed (approved):
Code review ACK 95a0104f2e9869799db84add108ae8c57b56d360. Looks good! Thanks for all your work on this breaking the changes down and making them simple.

Since last review, more log newlines are removed, -norpccookiefile code is reorganized a little without changing behavior, combine_logs.py assert is improved, feature_config_args.py more consistently stops nodes after tests, and using -noconf now triggers an info log instead of a warning.

l0rinc reviewed (commented):
I find a lot of negations very confusing, but I guess it's existing functionality, so we might as well embrace it

ryanofsky commented on src/bitcoin-util.cpp:
```diff
@@ -50,11 +50,11 @@ static int AppInitUtil(ArgsManager& args, int argc, char* argv[])
     }
 
-    if (HelpRequested(args) || args.IsArgSet("-version")) {
+    if (HelpRequested(args) || args.GetBoolArg("-version", false)) {
```
The version number is always going to be displayed whether you run `bitcoind` with `-version` `-help` or no arguments at all, because it is logged at startup. If you don't want to see the version number you need to run `bitcoind` with `-noprinttoconsole` or `-daemon` to not log to the console.

The point of `-version` is to show the version number and copyright information and exit, and the point of `-noversion` is to do the opposite and start up normally. You can think of `-version` as being short for `-print-version-then-exit=1` and `-noversion` as being short for `-print-version-then-exit=0`. I doubt there are any realistic use-cases for `-noversion`, and new behavior seems less confusing than previous behavior. The point of the commit is just to make this option behave like other boolean options, where `-nooption` does the opposite of `-option` instead of doing the same thing.

l0rinc replied:
I'm not yet sure this makes the behavior clearer for the version parameter, but the rest seems fine to me, thanks for clarifying.

hodlinator replied:
Thanks @ryanofsky for arguing my case. One could possibly stretch `-noversion` to actually remove the version from the default log, but not sure if that would be more useful. The fact that one of `-version`/`-noversion` exits quickly is mildly confusing.

---

> Tried:
> 
> ```python
>         self.log.info("Start bitcoin with -noversion for version information")
>         self.nodes[0].start(extra_args=['-version=0']) # -version=1 works
>         # Node should exit immediately and output version to stdout.
>         output, _ = self.get_node_output(ret_code_expected=0)
>         assert b'version' in output
>         self.log.info(f"Version text received: {output[0:60]} (...)")
> ```
> 
> which just hangs with no RPC connection available.

That is slightly incorrect. Here is the full error I got when trying to reproduce your change:

```
2024-12-04T15:35:30.230000Z TestFramework (INFO): PRNG seed is: 6906262290736093521
2024-12-04T15:35:30.230000Z TestFramework (DEBUG): Setting up network thread
2024-12-04T15:35:30.231000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_n5c4b731
2024-12-04T15:35:30.231000Z TestFramework (INFO): Start bitcoin with -h for help text
2024-12-04T15:35:30.232000Z TestFramework.node0 (DEBUG): bitcoind started, waiting for RPC to come up
2024-12-04T15:35:30.239000Z TestFramework (INFO): Help text received: b'Bitcoin Core daemon version v28.99.0-95a0104f2e98-dirty\n\nThe' (...)
2024-12-04T15:35:30.239000Z TestFramework (INFO): Start bitcoin with -version for version information
2024-12-04T15:35:30.240000Z TestFramework.node0 (DEBUG): bitcoind started, waiting for RPC to come up
2024-12-04T15:36:30.240000Z TestFramework (ERROR): Unexpected exception caught during testing
Traceback (most recent call last):
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_framework.py", line 132, in main
    self.run_test()
  File "/home/hodlinator/bitcoin/build/test/functional/feature_help.py", line 48, in run_test
    output, _ = self.get_node_output(ret_code_expected=0)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/hodlinator/bitcoin/build/test/functional/feature_help.py", line 20, in get_node_output
    ret_code = self.nodes[0].process.wait(timeout=60)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/wfbjq35kxs6x83c3ncpfxdyl5gbhdx4h-python3-3.12.6/lib/python3.12/subprocess.py", line 1264, in wait
    return self._wait(timeout=timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/wfbjq35kxs6x83c3ncpfxdyl5gbhdx4h-python3-3.12.6/lib/python3.12/subprocess.py", line 2045, in _wait
    raise TimeoutExpired(self.args, timeout)
subprocess.TimeoutExpired: Command '['/home/hodlinator/bitcoin/build/src/bitcoind', '-datadir=/tmp/bitcoin_func_test_n5c4b731/node0', '-logtimemicros', '-debug', '-debugexclude=libevent', '-debugexclude=leveldb', '-debugexclude=rand', '-uacomment=testnode0', '-disablewallet', '-logthreadnames', '-logsourcelocations', '-loglevel=trace', '-v2transport=0', '-version=0']' timed out after 60 seconds
2024-12-04T15:36:30.245000Z TestFramework (DEBUG): Closing down network thread
2024-12-04T15:36:30.296000Z TestFramework (INFO): Stopping nodes
2024-12-04T15:36:30.296000Z TestFramework.node0 (DEBUG): Stopping node
Traceback (most recent call last):
  File "/home/hodlinator/bitcoin/build/test/functional/feature_help.py", line 62, in <module>
    HelpTest(__file__).main()
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_framework.py", line 155, in main
    exit_code = self.shutdown()
                ^^^^^^^^^^^^^^^
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_framework.py", line 318, in shutdown
    self.stop_nodes()
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_framework.py", line 587, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_node.py", line 397, in stop_node
    self.stop(wait=wait)
    ^^^^^^^^^
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_node.py", line 215, in __getattr__
    assert self.rpc_connected and self.rpc is not None, self._node_msg("Error: no RPC connection")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: [node 0] Error: no RPC connection
[node 0] Cleaning up leftover process
```

It's the call to `process.wait()` in the local `get_node_output()` in *feature_help.py* that times out. The RPC assertion failure is just a knock-on issue that is fixed by #30660.

<details>
<summary>
Log with #30660 cherry-picked on top.
</summary>

```
2024-12-04T15:41:34.993000Z TestFramework (INFO): PRNG seed is: 7034279536394037572
2024-12-04T15:41:34.994000Z TestFramework (DEBUG): Setting up network thread
2024-12-04T15:41:34.994000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_3wqbq2m9
2024-12-04T15:41:34.994000Z TestFramework (INFO): Start bitcoin with -h for help text
2024-12-04T15:41:34.995000Z TestFramework.node0 (DEBUG): bitcoind started with PID 881245, waiting for RPC to come up
2024-12-04T15:41:35.002000Z TestFramework (INFO): Help text received: b'Bitcoin Core daemon version v28.99.0-95a0104f2e98-dirty\n\nThe' (...)
2024-12-04T15:41:35.002000Z TestFramework (INFO): Start bitcoin with -version for version information
2024-12-04T15:41:35.003000Z TestFramework.node0 (DEBUG): bitcoind started with PID 881246, waiting for RPC to come up
2024-12-04T15:42:35.004000Z TestFramework (ERROR): Unexpected exception caught during testing
Traceback (most recent call last):
  File "/home/hodlinator/bitcoin/test/functional/test_framework/test_framework.py", line 132, in main
    self.run_test()
  File "/home/hodlinator/bitcoin/build/test/functional/feature_help.py", line 48, in run_test
    output, _ = self.get_node_output(ret_code_expected=0)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/hodlinator/bitcoin/build/test/functional/feature_help.py", line 20, in get_node_output
    ret_code = self.nodes[0].process.wait(timeout=60)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/wfbjq35kxs6x83c3ncpfxdyl5gbhdx4h-python3-3.12.6/lib/python3.12/subprocess.py", line 1264, in wait
    return self._wait(timeout=timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/wfbjq35kxs6x83c3ncpfxdyl5gbhdx4h-python3-3.12.6/lib/python3.12/subprocess.py", line 2045, in _wait
    raise TimeoutExpired(self.args, timeout)
subprocess.TimeoutExpired: Command '['/home/hodlinator/bitcoin/build/src/bitcoind', '-datadir=/tmp/bitcoin_func_test_3wqbq2m9/node0', '-logtimemicros', '-debug', '-debugexclude=libevent', '-debugexclude=leveldb', '-debugexclude=rand', '-uacomment=testnode0', '-disablewallet', '-logthreadnames', '-logsourcelocations', '-loglevel=trace', '-v2transport=0', '-version=0']' timed out after 60 seconds
2024-12-04T15:42:35.009000Z TestFramework (DEBUG): Closing down network thread
2024-12-04T15:42:35.060000Z TestFramework (INFO): Stopping nodes
2024-12-04T15:42:35.060000Z TestFramework.node0 (DEBUG): Stopping node
2024-12-04T15:42:35.060000Z TestFramework.node0 (WARNING): Cannot call stop-RPC as we are not connected. Killing process 881246 so that wait_until_stopped will not time out.
2024-12-04T15:42:35.111000Z TestFramework.node0 (DEBUG): Node stopped
2024-12-04T15:42:35.111000Z TestFramework (WARNING): Not cleaning up dir /tmp/bitcoin_func_test_3wqbq2m9
2024-12-04T15:42:35.111000Z TestFramework (ERROR): Test failed. Test logging available at /tmp/bitcoin_func_test_3wqbq2m9/test_framework.log
2024-12-04T15:42:35.111000Z TestFramework (ERROR): 
2024-12-04T15:42:35.112000Z TestFramework (ERROR): Hint: Call /home/hodlinator/bitcoin/test/functional/combine_logs.py '/tmp/bitcoin_func_test_3wqbq2m9' to consolidate all logs
2024-12-04T15:42:35.112000Z TestFramework (ERROR): 
2024-12-04T15:42:35.112000Z TestFramework (ERROR): If this failure happened unexpectedly or intermittently, please file a bug and provide a link or upload of the combined log.
2024-12-04T15:42:35.112000Z TestFramework (ERROR): https://github.com/bitcoin/bitcoin/issues
2024-12-04T15:42:35.112000Z TestFramework (ERROR): 
```

</details>

l0rinc commented:
utACK 95a0104f2e9869799db84add108ae8c57b56d360
The code seems better than before, though most of these negations seem confusing to me - but it wasn't introduced here.
I have tested the negation of the parameters and reviewed most of the code - but couldn't meaningfully opine on the cookie related new tests, I will leave that to other reviewers.

hodlinator commented on test/functional/combine_logs.py:
```diff
@@ -79,11 +79,12 @@ def read_logs(tmp_dir):
+    # Find out what the folder is called that holds node 0's debug.log file
+    debug_logs = list(pathlib.Path(tmp_dir).glob('node0/**/debug.log'))
+    if len(debug_logs) > 0:
+        assert len(debug_logs) < 2, 'Max one debug.log is supported, ' \
```
Sweet! Will do something like this if I need to retouch for other reasons. (Might use `RuntimeError`).

hodlinator replied:
(Resolved by now merged #31433).

achow101 commented:
ACK 95a0104f2e9869799db84add108ae8c57b56d360

vasild commented:
Further idea:

`-noconnect=0` is interpreted as `-connect=1` which is interpreted as `-connect=0.0.0.1`

hodlinator commented:
> Further idea:
> 
> `-noconnect=0` is interpreted as `-connect=1` which is interpreted as `-connect=0.0.0.1`

:sweat_smile: Good one! I'll create a "Good first issue" for it.

hodlinator commented:
Done! #31426. @vasild let me know what you think of the issue description and proposed solution.
